catalog/async/catalog-async-data/src/test/groovy/org/codice/ddf/catalog/async/data/LazyProcessResourceImplTest.groovy

```diff
@@ -0,0 +1,223 @@
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.catalog.async.data
+
+import ddf.catalog.resource.Resource
+import org.codice.ddf.catalog.async.data.impl.LazyProcessResourceImpl
+import spock.lang.Shared
+import spock.lang.Specification
+
+import java.util.function.Supplier
+
+import static org.codice.ddf.catalog.async.data.impl.ProcessResourceImpl.*
+
+class LazyProcessResourceImplTest extends Specification {
+
+    @Shared
+    String RESOURCE_NAME = "resourceName"
+
+    @Shared
+    int RESOURCE_SIZE = 1
+
+    @Shared
+    InputStream RESOURCE_INPUTSTREAM = Mock(InputStream)
+
+    @Shared
+    String RESOURCE_MIMETYPE = "mimeType"
+
+    @Shared
+    String METACARD_ID = "metacardId"
+
+    @Shared
+    URI RESOURCE_URI = new URI(CONTENT_SCHEME, METACARD_ID, null)
+
+    @Shared
+    String RESOURCE_QUALIFIER = ""
+
+
+    private Resource resource
+    private Supplier<Resource> supplier
+
+    def setup() {
+        resource = Mock(Resource) {
+            getSize() >> RESOURCE_SIZE
+            getInputStream() >> RESOURCE_INPUTSTREAM
+            getName() >> RESOURCE_NAME
+            getMimeTypeValue() >> RESOURCE_MIMETYPE
+        }
+
+        supplier = Mock(Supplier) {
+            get() >> resource
+        }
+    }
+
+
+    def 'new resource IllegalArgument on bad inputs'() {
+        when:
+        new LazyProcessResourceImpl(metacardId, resourceSupplier)
+
+        then:
+        thrown(IllegalArgumentException)
+
+        where:
+        metacardId   | resourceSupplier
+        null         | Mock(Supplier)
+        ""           | Mock(Supplier)
+        " "          | Mock(Supplier)
+        METACARD_ID | null
+
+    }
+
+    def 'new resource verifying lazy loading not loaded'() {
+        when:
+        new LazyProcessResourceImpl(METACARD_ID, supplier)
+
+        then:
+        0 * supplier.get()
+    }
+
+    def 'verify a new resource is lazy loaded at first get of resource specific field'() {
+        given:
+        Supplier<Resource> supplier = Mock(Supplier) {
+            1 * get() >> resource
+        }
+
+        long expectedSize = 3914
+        String uriString = "something:3839ab393df930303#frag"
+        URI expectedUri = new URI(uriString);
+        String expectedQualifier = "frag"
+
+        def lazyProcessResource = new LazyProcessResourceImpl(METACARD_ID, supplier)
+        lazyProcessResource.setSize(expectedSize)
+        lazyProcessResource.setUri(expectedUri)
+
+
+
+        when:
+        lazyProcessResource.getName()
+        lazyProcessResource.getMimeType()
+        lazyProcessResource.getName()
+        lazyProcessResource.close()
```

Comment:
This `when block` is a little confusing to me. It seems like we'd like 3 separate tests instead:

- `'test resource is loaded when getName'`
- `'test resource is loaded when getMimeType'`
- `'test resource is loaded when getInputStream'`

because any `get`- method should attempt to lead the resource.

Also, why is `close()` called here?

Reply:
Updated test. 

It was just testing that the resource would only get loaded once on multiple calls to getters that load the resource.

Reply:
Oh. I misunderstood the intention of this method. Maybe four unit tests are needed then:
- 'test resource is loaded when getName'
- 'test resource is loaded when getMimeType'
- 'test resource is loaded when getInputStream'
- 'test resource is loaded only once when multiple calls to getters'

Reply:
Split into the four unit tests.